libs/wai-utilities/src/Network/Wai/Utilities/Server.hs

```diff
@@ -185,21 +168,43 @@ measureRequests m rtree = withPathTemplate rtree $ \p ->
       requestCounter m p . duration 30 12 m p
 {-# INLINEABLE measureRequests #-}
 
+-- | Run both 'catchErrorsException', 'catchErrorsResponse'.
+catchErrors  :: Logger -> OnErrorMetrics -> Middleware
+catchErrors l m = catchErrorsException l m . catchErrorsResponse l
+{-# INLINEABLE catchErrors #-}
+
 -- | Create a middleware that catches exceptions and turns
 -- them into appropriate 'Error' responses, thereby logging
 -- as well as counting server errors (i.e. exceptions that
 -- yield 5xx responses).
-catchErrors :: Logger -> Metrics -> Middleware
-catchErrors l m app req k =
+--
+-- This does not log any 'Response' values with error status.
+-- See 'catchErrors'.
+catchErrorsException :: Logger -> OnErrorMetrics -> Middleware
+catchErrorsException l m app req k =
     app req k `catch` errorResponse
   where
     errorResponse ex = do
         er <- runHandlers ex errorHandlers
         when (statusCode (Error.code er) >= 500) $
-            logIO l Log.Error (Just req) (oneline <$> show ex)
+            logIO l Log.Error (Just req) (show ex)
         onError l m req k er
-    oneline c = if isSpace c then ' ' else c
-{-# INLINEABLE catchErrors #-}
+{-# INLINEABLE catchErrorsException #-}
+
+-- | Catch status >= 400 even if it does not come in the form of an exception, but as a
+-- response, and log a warning that this really shouldn't happen, and a typed exception should
+-- be thrown instead.
+catchErrorsResponse :: Logger -> Middleware
+catchErrorsResponse l app req k =
+    app req $ \resp -> do
+        when (statusCode (responseStatus resp) >= 400) $ do
+            let errinfo = Error (responseStatus resp) (cs . statusMessage $ responseStatus resp) "N/A"
+            Log.warn l
+                $ field "request" (fromMaybe "N/A" (lookupRequestId req))
+                . field "error" (show errinfo)
+                . (msg $ val "Expected exception instead of response with status >= 400")
```

Comment:
```suggestion
                . (msg $ val "A handler responded with status >= 400; please throw an exception instead of manually constructing an error response")
```